### PR TITLE
provider/datadog: Add datadog_downtime resource

### PR DESCRIPTION
**Pull request by bflad**
Hello! This should hopefully get the ball rolling for adding a `datadog_downtime` resource to the datadog provider, which is used to manage alerting downtimes. References:
* [Datadog Monitoring Guide](http://docs.datadoghq.com/guides/monitoring/#downtime)
* [Datadog UI](https://app.datadoghq.com/monitors#/downtime)
* [Datadog API Reference](http://docs.datadoghq.com/api/?lang=python#downtimes)
* [go-datadog-api Reference](https://godoc.org/github.com/zorkian/go-datadog-api#Client.CreateDowntime)

This appears to be pretty close for an initial implementation, however I am having some issues working with the Terraform schema and Go (first time writing in Go, so apologies in advance!). I used the existing `datadog_monitor` as a reference for writing this resource.

Known Working:
* CRUD operations for downtimes
* Verify "deletion" in testing by checking `Active` attribute (Datadog doesn't actually delete the downtime -- it does disappear from the UI in this state though)
* Unit and acceptance testing passes

Known Issues:
* ~~The schema for `recurrence.week_days` should be a validated List of week days -- currently throws `	* '[week_days]' expected type 'string', got unconvertible type '[]interface {}'` when defined in configuration~~
* ~~The `recurrence` state is occasionally unstable for certain attributes (mostly `period`), but I'm very unsure why (Go pointers?). Here's an example that testing will not break every run, but almost always in same the way:~~

```
--- FAIL: TestAccDatadogDowntime_BasicUntilDateRecurrence (1.04s)
	testing.go:265: Step 0 error: After applying this step and refreshing, the plan was not empty:

		DIFF:

		UPDATE: datadog_downtime.foo
		  recurrence.%:          "1" => "3"
		  recurrence.period:     "" => "1"
		  recurrence.until_date: "" => "1736226000"

		STATE:

		datadog_downtime.foo:
		  ID = 207794591
		  active = false
		  disabled = false
		  end = 1735765200
		  message = Example Datadog downtime message.
		  recurrence.% = 1
		  recurrence.type = days
		  scope.# = 1
		  scope.0 = host:UntilDateRecurrence
		  start = 1735707600
```

Any help or feedback is appreciated, thanks!

### CODE REVIEW

File: builtin/providers/datadog/resource_datadog_downtime.go [hunk 1]
```diff
@@ -0,0 +1,306 @@
+package datadog
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/zorkian/go-datadog-api"
```
Comment (ojongerius) [inline builtin/providers/datadog/resource_datadog_downtime.go]: Needs to be updated to "gopkg.in/zorkian/go-datadog-api.v2"

File: builtin/providers/datadog/resource_datadog_downtime.go [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+							Type:          schema.TypeInt,
+							Optional:      true,
+							ConflictsWith: []string{"recurrence.until_date"},
+						},
+						"week_days": {
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem: &schema.Schema{
+								Type:         schema.TypeString,
+								ValidateFunc: validateWeekDay,
```
Comment (ojongerius) [inline builtin/providers/datadog/resource_datadog_downtime.go]: Optional, but validateWeekDay be defined here, see this example: https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_kinesis_firehose_delivery_stream.go#L76:

```go
	ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
		value := v.(string)
		if value != "s3" && value != "redshift" && value != "elasticsearch" {
			errors = append(errors, fmt.Errorf(
				"%q must be one of 's3', 'redshift', 'elasticsearch'", k))
		}
		return
	},
```

File: builtin/providers/datadog/resource_datadog_downtime.go [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+	}
+	return []*schema.ResourceData{d}, nil
+}
+
+// validateWeekDay ensures that the week_days resource parameter is
+// correct.
+func validateWeekDay(v interface{}, k string) (ws []string, errors []error) {
+	value := v.(string)
+
+	if value != "Mon" && value != "Tue" && value != "Wed" && value != "Thu" && value != "Fri" && value != "Sat" && value != "Sun" {
```
Comment (ojongerius) [inline builtin/providers/datadog/resource_datadog_downtime.go]: I would use switch here, see https://gobyexample.com/switch

Something like (untested):

```go
	switch value {
	case "Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun":
		errors = append(errors, fmt.Errorf(
		"%q contains an invalid week day parameter %q. Valid parameters are Mon, Tue, Wed, Thu, Fri, Sat or Sun",
		k))
	}
	return
```

**Comment by ojongerius** [comment]
> "The schema for recurrence.week_days should be a validated List of week days -- currently throws * '[week_days]' expected type 'string', got unconvertible type '[]interface {}' when defined in configuration"

This is thrown while planning during your test, it is triggered by using a list nested in a map I see it in `TestAccDatadogDowntime_WeekDayRecurring`:

```
--- FAIL: TestAccDatadogDowntime_WeekDayRecurring (0.29s)
        testing.go:265: Step 0 error: Error planning: 1 error(s) occurred:
                
                * datadog_downtime.foo: 1 error(s) occurred:
                
                * datadog_downtime.foo: recurrence: 1 error(s) decoding:
                
                * '[week_days]' expected type 'string', got unconvertible type '[]interface {}'
=== RUN   TestAccDatadogMonitor_Basic

```

Which contains:

```hcl
  recurrence {
    period    = 1
	type      = "weeks"
	week_days = ["Sat", "Sun"]
  }
```

The error is not in your code, but the schema, `schema.TypeList` for "week_days" nested in `schema.TypeMap` ("recurrence"). I think this *should* work, but it does not, at the moment.

Terraform does not catch this for you, and the error message is not very informative. AFAICS you'll have to work around this, or fix this in Terraform core. Don't have a good solution for you at the moment, but will let you know if I think of something.

You'll see you'll be able to use `week_days` if you move it up into the top of your resource, but I understand this not a very elegant solution.

**Comment by bflad** [comment]
@ojongerius thanks for the feedback! I spent some time with this today and got it working as expected (hopefully the right way 😄).

PR updates:
* Switch PR base to 0-8-stable
* Upgrade datadog_dowtime library to v2
* Switch recurrence schema from TypeMap to TypeList -- this makes the nested TypeList work interestingly enough
* Add recurrence type validation function
* Add testing for validation functions

Here's what I'm seeing now:
```
make test TEST=./builtin/providers/datadog TESTARGS='-run=TestResourceDatadogDowntime'
==> Checking that code complies with gofmt requirements...
==> Checking AWS provider for unchecked errors...
==> NOTE: at this time we only look for uncheck errors in the AWS package
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/26 17:54:55 Generated command/internal_plugin_list.go
TF_ACC= go test ./builtin/providers/datadog -run=TestResourceDatadogDowntime -timeout=30s -parallel=4
ok  	github.com/hashicorp/terraform/builtin/providers/datadog	0.014s

make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestAccDatadogDowntime'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/26 17:55:12 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestAccDatadogDowntime -timeout 120m
=== RUN   TestAccDatadogDowntime_Basic
--- PASS: TestAccDatadogDowntime_Basic (1.93s)
=== RUN   TestAccDatadogDowntime_BasicMultiScope
--- PASS: TestAccDatadogDowntime_BasicMultiScope (1.76s)
=== RUN   TestAccDatadogDowntime_BasicNoRecurrence
--- PASS: TestAccDatadogDowntime_BasicNoRecurrence (1.50s)
=== RUN   TestAccDatadogDowntime_BasicUntilDateRecurrence
--- PASS: TestAccDatadogDowntime_BasicUntilDateRecurrence (1.39s)
=== RUN   TestAccDatadogDowntime_BasicUntilOccurrencesRecurrence
--- PASS: TestAccDatadogDowntime_BasicUntilOccurrencesRecurrence (1.15s)
=== RUN   TestAccDatadogDowntime_WeekDayRecurring
--- PASS: TestAccDatadogDowntime_WeekDayRecurring (1.26s)
=== RUN   TestAccDatadogDowntime_Updated
--- PASS: TestAccDatadogDowntime_Updated (2.47s)
=== RUN   TestAccDatadogDowntime_TrimWhitespace
--- PASS: TestAccDatadogDowntime_TrimWhitespace (1.18s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/datadog	12.661s
```

**Comment by ojongerius** [comment]
Nice work!

Looks like the last hurdle is getting `TestDatadogDowntime_import` to pass:

```
▶ make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestDatadogDowntime'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/27 15:22:52 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestDatadogDowntime -timeout 120m
=== RUN   TestDatadogDowntime_import
--- FAIL: TestDatadogDowntime_import (2.48s)
        testing.go:265: Step 0 error: Error applying: 1 error(s) occurred:
                
                * datadog_downtime.foo: error updating downtime: API error 400 Bad Request: {"errors":["Scheduled downtime start cannot be in the past"]}
FAIL
exit status 1
FAIL    github.com/hashicorp/terraform/builtin/providers/datadog        2.490s
make: *** [testacc] Error 1
```

File: builtin/providers/datadog/import_datadog_downtime_test.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+const testAccCheckDatadogDowntimeConfigImported = `
+resource "datadog_downtime" "foo" {
+  scope = ["host:X", "host:Y"]
+  start = 1483308000
```
Comment (ojongerius) [inline builtin/providers/datadog/import_datadog_downtime_test.go]: These need to be in the future, get fancy and generate them dynamically or plan further ahead ;)

**Comment by bflad** [comment]
I've adjusted the import test timestamps to be similar to the other tests in the future and its working locally for me:
```
make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestDatadogDowntime'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/27 08:08:49 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestDatadogDowntime -timeout 120m
=== RUN   TestDatadogDowntime_import
--- PASS: TestDatadogDowntime_import (1.67s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/datadog	1.684s
```

Please let me know if there are other adjustments you'd like!

**Comment by ojongerius** [comment]
```
▶ make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestDatadogDowntime'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/28 09:11:00 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestDatadogDowntime -timeout 120m
=== RUN   TestDatadogDowntime_import
--- PASS: TestDatadogDowntime_import (6.64s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/datadog        6.656s

github.com/hashicorp/terraform  datadog_downtime_resource ✔                                                                                                                                                                                                                8h59m  
▶ make testacc TEST=./builtin/providers/datadog TESTARGS='-run=TestAccDatadogDowntime'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/02/28 09:11:19 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/datadog -v -run=TestAccDatadogDowntime -timeout 120m
=== RUN   TestAccDatadogDowntime_Basic
--- PASS: TestAccDatadogDowntime_Basic (5.87s)
=== RUN   TestAccDatadogDowntime_BasicMultiScope
--- PASS: TestAccDatadogDowntime_BasicMultiScope (4.66s)
=== RUN   TestAccDatadogDowntime_BasicNoRecurrence
--- PASS: TestAccDatadogDowntime_BasicNoRecurrence (4.94s)
=== RUN   TestAccDatadogDowntime_BasicUntilDateRecurrence
--- PASS: TestAccDatadogDowntime_BasicUntilDateRecurrence (5.06s)
=== RUN   TestAccDatadogDowntime_BasicUntilOccurrencesRecurrence
--- PASS: TestAccDatadogDowntime_BasicUntilOccurrencesRecurrence (5.19s)
=== RUN   TestAccDatadogDowntime_WeekDayRecurring
--- PASS: TestAccDatadogDowntime_WeekDayRecurring (4.92s)
=== RUN   TestAccDatadogDowntime_Updated
--- PASS: TestAccDatadogDowntime_Updated (10.06s)
=== RUN   TestAccDatadogDowntime_TrimWhitespace
--- PASS: TestAccDatadogDowntime_TrimWhitespace (4.74s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/datadog        45.465s
```

**Comment by stack72** [comment]
LGTM Thanks for all the work here @bflad and for the review @ojongerius

**Pull request merged**